File: typescript/src/part3.ts
import readline from "readline";

const RST  = "\x1b[0m";
const BOLD = "\x1b[1m";
const DIM  = "\x1b[2m";
const RED  = "\x1b[91m";
const GRN  = "\x1b[92m";
const YLW  = "\x1b[93m";
const BLU  = "\x1b[94m";
const MAG  = "\x1b[95m";
const CYN  = "\x1b[96m";
const GRY  = "\x1b[90m";

function codeOpen(lang: string, col: string) {
  console.log(`${DIM}  ╭─ ${RST}${BOLD}${col}${lang}${RST} ${DIM}────────────────────────────────────────────────────`);
  console.log(`  │${RST}`);
}
const codeOpenTs = () => codeOpen("typescript", BLU);
function codeClose() {
  console.log(`${DIM}  │`);
  console.log(`  ╰──────────────────────────────────────────────────────────${RST}\n`);
}
function sectionResult() {
  console.log(`${BOLD}${CYN}\n  ━━ Result ${RST}${DIM}━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━${RST}\n`);
}
function sectionExplain() {
  console.log(`${BOLD}${MAG}\n  ━━ What happened ${RST}${DIM}━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━${RST}\n`);
}

const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
let rlClosed = false;
rl.on("close", () => { rlClosed = true; });

async function pressEnter(question: string): Promise<void> {
  console.log(`${BOLD}${YLW}\n  ╔═ ? ════════════════════════════════════════════════════════╗${RST}`);
  console.log(`${BOLD}${YLW}  ║${RST}  ${question}`);
  console.log(`${BOLD}${YLW}  ║${RST}`);
  console.log(`${BOLD}${YLW}  ║${RST}  ${DIM}Press Enter to continue...${RST}`);
  console.log(`${BOLD}${YLW}  ╚════════════════════════════════════════════════════════════╝${RST}`);
  if (rlClosed) return;
  await new Promise<void>(resolve => rl.once("line", () => { console.log(); resolve(); }));
}

const nsNow = (): bigint => process.hrtime.bigint();
const p = () => `  ${DIM}│${RST}  `;

// ══════════════════════════════════════════════════════════════════
// Demo 1 — shared ownership (verbosity)
// ══════════════════════════════════════════════════════════════════
async function demoSharedOwnership(): Promise<void> {
  const pp = p();

  console.log(`  In TypeScript, sharing is just ${BLU}assignment${RST}. The GC traces all`);
  console.log(`  live references and frees memory in batches when it decides to:\n`);

  codeOpenTs();
  console.log(`${pp}${YLW}interface${RST} ${GRN}Node${RST} { value: ${YLW}number${RST} }`);
  console.log(`${DIM}  │${RST}`);
  console.log(`${pp}${YLW}const${RST} a: ${GRN}Node${RST} = { value: ${MAG}42${RST} };   ${GRY}// heap alloc — GC owns it${RST}`);
  console.log(`${pp}${YLW}const${RST} b = a;                 ${GRY}// aliases the same object${RST}`);
  console.log(`${pp}console.log(a.value, b.value); ${GRY}// both work${RST}`);
  console.log(`${pp}${GRY}// GC frees when no refs remain — timing is not your call${RST}`);
  codeClose();

  await pressEnter("Zero boilerplate — but what does TypeScript actually give you here?");
  sectionResult();

  interface Node { value: number }
  const a: Node = { value: 42 };
  const b = a;
  console.log(`  a.value=${GRN}${BOLD}${a.value}${RST}  b.value=${GRN}${BOLD}${b.value}${RST}  (same object in memory)`);
  b.value = 99;
  console.log(`  After b.value = 99: a.value=${YLW}${BOLD}${a.value}${RST}  ← a sees the mutation\n`);

  console.log(`  ${BOLD}Lines of boilerplate to safely share one heap value:\n${RST}`);
  console.log(`  ${RED}C (manual refcount)${RST}  ~15 lines  ${RED}███████████████${RST}`);
  console.log(`  ${GRN}Rust (Rc<T>)${RST}          4 lines  ${GRN}████${RST}`);
  console.log(`  ${BLU}TypeScript (GC)${RST}        0 lines  ${DIM}(GC tracks references invisibly)${RST}`);

  sectionExplain();
  console.log(`  TypeScript has no ownership — ${BLU}b = a${RST} is ${BOLD}aliasing${RST}, not a ref-count.`);
  console.log(`  The GC traces all reachable objects and frees them in batches.\n`);
  console.log(`  The ${YLW}trade-off${RST}:`);
  console.log(`    ${GRN}•${RST} Zero verbosity — sharing is natural and invisible`);
  console.log(`    ${RED}•${RST} Mutations through any alias affect all holders (shown above)`);
  console.log(`    ${RED}•${RST} No control over ${YLW}when${RST} memory is freed — see demo 2\n`);
  console.log(`  ${GRY}In Rust, Rc<T> gives shared ${BOLD}immutable${GRY} access; mutation`);
  console.log(`  ${GRY}requires explicit ${BOLD}RefCell<T>${GRY} — the compiler enforces the distinction.${RST}`);
}

// ══════════════════════════════════════════════════════════════════
// Demo 2 — allocation latency benchmark
// ══════════════════════════════════════════════════════════════════
const BATCH_SIZE  = 1000;
const BATCH_COUNT = 500;

interface Node { a: number; b: number; c: number; d: number; e: number; f: number }

async function demoAllocLatency(): Promise<void> {
  const pp = p();

  console.log(`  Each batch: ${YLW}${BATCH_SIZE}${RST} object allocations — the GC collects them later.`);
  console.log(`  Total: ${YLW}500,000${RST} allocations across ${YLW}${BATCH_COUNT}${RST} batches.\n`);

  codeOpenTs();
  console.log(`${pp}${YLW}const${RST} nodes: ${GRN}Node[]${RST} = ${YLW}new${RST} Array(BATCH_SIZE);`);
  console.log(`${pp}${YLW}for${RST} (${YLW}let${RST} i = 0; i < BATCH_SIZE; i++)`);
  console.log(`${pp}    nodes[i] = { a: i, b: i*2, ${GRY}/* ... */${RST} };`);
  console.log(`${pp}${GRY}// nodes goes out of scope — GC frees it when it decides to${RST}`);
  codeClose();

  await pressEnter("The GC accumulates garbage between batches. Will p95 differ from p50?");
  sectionResult();

  const times: number[] = new Array(BATCH_COUNT);
  let checksum = 0;

  for (let b = 0; b < BATCH_COUNT; b++) {
    const filled = Math.floor((b + 1) * 38 / BATCH_COUNT);
    const bar = GRN + "▓".repeat(filled) + RST + DIM + "░".repeat(38 - filled) + RST;
    process.stdout.write(`\r  ${DIM}[${RST}${bar}${DIM}]${RST}  ${b + 1}/${BATCH_COUNT}   `);

    const t0 = nsNow();
    const nodes: Node[] = new Array(BATCH_SIZE);
    for (let i = 0; i < BATCH_SIZE; i++) {
      nodes[i] = { a: i, b: i*2, c: i*3, d: i*4, e: i*5, f: i*6 };
    }
    for (const n of nodes) checksum += n.a;
    times[b] = Number(nsNow() - t0) / 1000; /* µs */
  }
  process.stdout.write(`\r  ${GRN}Complete.${RST}                                              \n\n`);

  times.sort((a, b) => a - b);
  const p50  = times[Math.floor(BATCH_COUNT / 2)];
  const p95  = times[Math.floor(BATCH_COUNT * 0.95)];
  const tmax = times[BATCH_COUNT - 1];
  const sc   = Math.max(tmax, 1);

  const bar = (n: number) => "█".repeat(Math.floor(n * 30 / sc));
  console.log(`  ${BOLD}Batch time (${BATCH_COUNT} batches × ${BATCH_SIZE} allocs):${RST}\n`);
  console.log(`   p50  ${GRN}${BOLD}${p50.toFixed(0).padStart(5)}${RST}µs  ${GRN}${bar(p50)}${RST}`);
  console.log(`   p95  ${YLW}${BOLD}${p95.toFixed(0).padStart(5)}${RST}µs  ${YLW}${bar(p95)}${RST}`);
  console.log(`   max  ${RED}${BOLD}${tmax.toFixed(0).padStart(5)}${RST}µs  ${RED}${bar(tmax)}${RST}`);
  console.log(`\n  ${DIM}(checksum=${checksum} — prevents dead-code elimination)${RST}`);

  sectionExplain();
  if (p95 > p50 * 2.5) {
    console.log(`  ${RED}${BOLD}GC pause visible:${RST} p95 is ${(p95 / p50).toFixed(1)}× p50.`);
    console.log(`  The GC swept accumulated garbage during at least one batch.\n`);
  } else {
    console.log(`  ${YLW}No large GC spike this run${RST} — JSC may have collected`);
    console.log(`  during your keypress pauses. Try running non-interactively:\n`);
    console.log(`  ${DIM}  printf '2\\n\\nq\\n' | bun run src/part3.ts${RST}\n`);
  }
  console.log(`  The GC runs ${YLW}non-deterministically${RST}. When it fires, it pauses`);
  console.log(`  the entire process — that spike appears in ${RED}p95 and max${RST}.\n`);
  console.log(`  C and Rust call free()/drop() ${BOLD}immediately${RST} per batch.`);
  console.log(`  No accumulation → ${GRN}p50 ≈ p95 ≈ max${RST}: flat, predictable latency.\n`);
  console.log(`  ${YLW}For most applications this doesn't matter.${RST}`);
  console.log(`  For game loops, real-time audio, or trading systems: it does.`);
}

// ─── menu ─────────────────────────────────────────────────────────────
async function printMenu(): Promise<string> {
  console.log(`${BLU}${BOLD}\n  ┌──────────────────────────────────────┐`);
  console.log(`  │ TypeScript — Performance & Verbosity │`);
  console.log(`  └──────────────────────────────────────┘${RST}`);
  console.log(`${DIM}  Part 3: memory management cost in code and latency.\n${RST}`);
  console.log(`  ${CYN}1)${RST} Shared ownership — verbosity comparison`);
  console.log(`  ${CYN}2)${RST} Allocation latency — deterministic free vs GC`);
  console.log(`  ${DIM}q) Quit${RST}`);
  process.stdout.write(`\n  ${BOLD}> ${RST}`);
  if (rlClosed) process.exit(0);
  return new Promise(resolve =>
    rl.once("line", line => { console.log(); resolve(line.trim()); })
  );
}

async function main() {
  for (;;) {
    const choice = await printMenu();
    if      (choice === "1") await demoSharedOwnership();
    else if (choice === "2") await demoAllocLatency();
    else if (choice === "q" || choice === "Q") {
      console.log(`${DIM}\n  Exiting. GC will handle cleanup whenever it feels like it.\n${RST}`);
      rl.close();
      break;
    } else console.log(`${RED}  Unknown option.${RST}`);
  }
}

main();

File: rust/src/bin/part3.rs
use std::hint::black_box;
use std::io::{self, BufRead, Write};
use std::rc::Rc;
use std::time::Instant;

const RST:  &str = "\x1b[0m";
const BOLD: &str = "\x1b[1m";
const DIM:  &str = "\x1b[2m";
const GRN:  &str = "\x1b[92m";
const YLW:  &str = "\x1b[93m";
const BLU:  &str = "\x1b[94m";
const MAG:  &str = "\x1b[95m";
const CYN:  &str = "\x1b[96m";
const GRY:  &str = "\x1b[90m";
const RED:  &str = "\x1b[91m";

fn code_open(lang: &str, color: &str) {
    println!("{DIM}  ╭─ {RST}{BOLD}{color}{lang}{RST} {DIM}────────────────────────────────────────────────────");
    println!("  │{RST}");
}
fn code_open_rust() { code_open("rust", GRN); }
fn code_close() {
    println!("{DIM}  │");
    println!("  ╰──────────────────────────────────────────────────────────{RST}");
    println!();
}
fn section_result() {
    println!("{BOLD}{CYN}\n  ━━ Result {RST}{DIM}━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━{RST}\n");
}
fn section_explain() {
    println!("{BOLD}{MAG}\n  ━━ What happened {RST}{DIM}━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━{RST}\n");
}
fn press_enter(question: &str) {
    println!("{BOLD}{YLW}\n  ╔═ ? ════════════════════════════════════════════════════════╗{RST}");
    println!("{BOLD}{YLW}  ║{RST}  {question}");
    println!("{BOLD}{YLW}  ║{RST}");
    println!("{BOLD}{YLW}  ║{RST}  {DIM}Press Enter to continue...{RST}");
    println!("{BOLD}{YLW}  ╚════════════════════════════════════════════════════════════╝{RST}");
    io::stdout().flush().unwrap();
    let mut buf = String::new();
    let n = io::stdin().lock().read_line(&mut buf).unwrap_or(0);
    if n == 0 { std::process::exit(0); }
    println!();
}
fn p() -> String { format!("  {DIM}│{RST}  ") }

// ══════════════════════════════════════════════════════════════════
// Demo 1 — shared ownership (verbosity)
// ══════════════════════════════════════════════════════════════════
fn demo_shared_ownership() {
    let p = p();

    println!("  {GRN}Rc<T>{RST} is Rust's reference-counted smart pointer —");
    println!("  the same mechanism as C's manual refcount, generated for you:\n");

    code_open_rust();
    println!("{p}{YLW}use{RST} std::rc::{GRN}Rc{RST};");
    println!("{DIM}  │{RST}");
    println!("{p}{YLW}let{RST} a = {GRN}Rc{RST}::new({MAG}42{RST});          {GRY}// alloc + refs=1{RST}");
    println!("{p}{YLW}let{RST} b = {GRN}Rc{RST}::clone(&a);      {GRY}// refs=2  (retain){RST}");
    println!("{p}println!(\"{{a}}, count={{}}\", {GRN}Rc{RST}::strong_count(&a));");
    println!("{p}{GRY}// a and b drop here — freed exactly once{RST}");
    code_close();

    press_enter("4 lines vs ~15 in C. Does it behave identically at runtime?");
    section_result();

    let a = Rc::new(42i32);
    let b = Rc::clone(&a);
    println!("  a={GRN}{BOLD}{a}{RST}  b={GRN}{BOLD}{b}{RST}  strong_count={GRN}{BOLD}{}{RST}",
             Rc::strong_count(&a));
    drop(b);
    println!("  Dropped b → count={YLW}{}{RST}", Rc::strong_count(&a));
    drop(a);
    println!("  Dropped a → count=0 — freed\n");

    println!("  {BOLD}Lines of boilerplate to safely share one heap value:\n{RST}");
    println!("  {RED}C (manual refcount){RST}  ~15 lines  {RED}███████████████{RST}");
    println!("  {GRN}Rust (Rc<T>){RST}          4 lines  {GRN}████{RST}");
    println!("  {BLU}TypeScript (GC){RST}        0 lines  {DIM}(GC tracks references invisibly){RST}");

    section_explain();
    println!("  {GRN}Rc<T>{RST} wraps your value with a heap-allocated reference counter.");
    println!("  {GRN}Rc::clone(){RST} increments it ({YLW}retain{RST}); {GRN}drop(){RST} decrements it ({YLW}release{RST}).");
    println!("  When count hits 0, {GRN}Drop::drop(){RST} is called — exactly once.\n");
    println!("  You write 4 lines. Rust writes the other 11 for you.");
    println!("  The {YLW}trade-off{RST}: sharing is {BOLD}explicit{RST} in the type ({GRN}Rc<T>{RST} vs {YLW}T{RST}).");
    println!("  {BLU}TypeScript{RST}: zero lines, but {YLW}when{RST} memory frees is not your decision.");
}

// ══════════════════════════════════════════════════════════════════
// Demo 2 — allocation latency benchmark
// ══════════════════════════════════════════════════════════════════
const BATCH_SIZE:  usize = 1000;
const BATCH_COUNT: usize = 500;

struct Node { a: i32, b: i32, c: i32, d: i32, e: i32, f: i32 }

fn demo_alloc_latency() {
    let p = p();

    println!("  Each batch: {YLW}{BATCH_SIZE}{RST} {GRN}Box::new(){RST} allocations, then drop.");
    println!("  Total: {YLW}500,000{RST} allocations across {YLW}{BATCH_COUNT}{RST} batches.\n");
    println!("  {YLW}For accurate numbers run with:{RST}  {DIM}cargo run --release --bin part3{RST}\n");

    code_open_rust();
    println!("{p}{YLW}let mut{RST} nodes: {GRN}Vec<Box<Node>>{RST} = {GRN}Vec{RST}::with_capacity(BATCH_SIZE);");
    println!("{p}{YLW}for{RST} i {YLW}in{RST} 0..BATCH_SIZE {{");
    println!("{p}    nodes.{CYN}push{RST}({GRN}Box{RST}::new({GRN}Node{RST} {{ a: i, {GRY}/* ... */{RST} }}));");
    println!("{p}}}");
    println!("{p}{GRY}// nodes drops here — all freed deterministically{RST}");
    code_close();

    press_enter("Box::new/drop uses the same allocator as C malloc/free. Will numbers match?");
    section_result();

    let mut times: Vec<u128> = Vec::with_capacity(BATCH_COUNT);
    let mut checksum: i64 = 0;

    for b in 0..BATCH_COUNT {
        let filled = (b + 1) * 38 / BATCH_COUNT;
        print!("\r  {DIM}[{RST}");
        for x in 0..38usize {
            print!("{}", if x < filled { format!("{GRN}▓{RST}") } else { format!("{DIM}░{RST}") });
        }
        print!("{DIM}]{RST}  {}/{BATCH_COUNT}   ", b + 1);
        io::stdout().flush().unwrap();

        let t0 = Instant::now();
        let mut nodes: Vec<Box<Node>> = Vec::with_capacity(BATCH_SIZE);
        for i in 0..BATCH_SIZE {
            let i = i as i32;
            nodes.push(black_box(Box::new(Node {
                a: i, b: i*2, c: i*3, d: i*4, e: i*5, f: i*6,
            })));
        }
        for n in &nodes { checksum += n.a as i64; }
        drop(nodes); /* deterministic free */
        times.push(t0.elapsed().as_nanos() / 1000); /* µs */
    }
    println!("\r  {GRN}Complete.{RST}                                              \n");

    times.sort_unstable();
    let p50  = times[BATCH_COUNT / 2];
    let p95  = times[BATCH_COUNT * 95 / 100];
    let tmax = *times.last().unwrap();
    let sc   = tmax.max(1);

    let bar = |n: u128| "█".repeat((n * 30 / sc) as usize);
    println!("  {BOLD}Batch time ({BATCH_COUNT} batches × {BATCH_SIZE} allocs):{RST}\n");
    println!("   p50  {GRN}{BOLD}{p50:>5}{RST}µs  {GRN}{}{RST}", bar(p50));
    println!("   p95  {YLW}{BOLD}{p95:>5}{RST}µs  {YLW}{}{RST}", bar(p95));
    println!("   max  {RED}{BOLD}{tmax:>5}{RST}µs  {RED}{}{RST}", bar(tmax));
    println!("\n  {DIM}(checksum={checksum} — prevents dead-code elimination){RST}");

    section_explain();
    println!("  {GRN}drop(){RST} is {BOLD}deterministic{RST} — it runs the instant nodes goes out of scope.");
    println!("  Rust uses the system allocator (same as C). Numbers should be close.\n");
    println!("  Compare with the TypeScript CLI part 3:");
    println!("    {YLW}If TS p95 >> p50{RST}: you're seeing a GC pause.");
    println!("    {GRN}If TS p95 ≈ p50{RST}:  JSC collected during your keypress gaps.");
    println!("                  Run non-interactively to remove that slack.\n");
    println!("  Neither C nor Rust can produce a GC spike — there is no GC.");
}

// ─── menu ────────────────────────────────────────────────────────────
fn print_menu() {
    println!("{GRN}{BOLD}\n  ┌──────────────────────────────────────┐");
    println!("  │   Rust — Performance & Verbosity     │");
    println!("  └──────────────────────────────────────┘{RST}");
    println!("{DIM}  Part 3: memory management cost in code and latency.\n{RST}");
    println!("  {CYN}1){RST} Shared ownership — verbosity comparison");
    println!("  {CYN}2){RST} Allocation latency — deterministic free vs GC");
    println!("  {DIM}q) Quit{RST}");
    print!("\n  {BOLD}> {RST}");
    io::stdout().flush().unwrap();
}

fn main() {
    let stdin = io::stdin();
    loop {
        print_menu();
        let mut line = String::new();
        let n = stdin.lock().read_line(&mut line).unwrap_or(0);
        if n == 0 { break; }
        println!();
        match line.trim() {
            "1" => demo_shared_ownership(),
            "2" => demo_alloc_latency(),
            "q" | "Q" => {
                println!("{DIM}\n  Exiting. All memory freed deterministically.\n{RST}");
                break;
            }
            _ => println!("{RED}  Unknown option.{RST}"),
        }
    }
}
